File: Algorithms/Examples/Part3/C/queue.c
#include <stdio.h>
#include <stdlib.h>
#include "queue.h"

Node* createNode(int data) {
	Node* n = malloc(sizeof(Node));
	n->data = data;
	n->next = NULL;
	return n;
}

Queue* createQueue() {
	Queue* q = malloc(sizeof(Queue));
	q->head = q->tail = NULL;
	return q;
}

int isEmpty(Queue* q) {
	return (q->tail == NULL);
}

void enqueue(Queue* q, int val) {

	Node* newnode = createNode(val);
	if (!isEmpty(q)) {
		q->tail->next = newnode;
		q->tail = newnode;	
	} else {
		q->head = q->tail = newnode;
	}
}

int dequeue(Queue* q) {
	if (!isEmpty(q)) {
		Node* n = q->head;
		int res = n->data;
		q->head = q->head->next;
		if (q->head == NULL)
			q->tail = NULL;
		free(n);
		return res;
	} else {
		return -1;
	}	
}

int first(Queue* q) {
	if (!isEmpty(q)) {
		return q->head->data;
	} else {
		return -1;
	}
}

File: Algorithms/Examples/Part3/Java/Queue.java
package Java;

public class Queue {

	private int head, tail, capacity;
	private int[] arr;

	public Queue(int capacity) {
		this.capacity = capacity;
		head = tail = 0;
		arr = new int[capacity];
	}

	public boolean isEmpty() {
		return (tail == 0);
	}

	public void enqueue(int x) {
		if (tail == capacity - 1) {
			System.out.println("Queue is full");
			return;
		}
		arr[++tail] = x;
	}

	public int dequeue() {
		if (!isEmpty()) {
			int res = arr[head];
			for (int i = 0; i < tail - 1; i++)
				arr[i] = arr[i + 1];
			if (tail < capacity)
				arr[tail--] = 0;
			return res;
		}
		return -1;	
	}

	public int first() {
		if (!isEmpty())
			return arr[head];
		else
			return -1;
	}

	public static void main(String[] args) {
		Queue q = new Queue(10);
		for (int i = 0; i < 12; i++)
			q.enqueue(i);
		for (int i = 0; i < 12; i++)
			System.out.println(q.dequeue() + ", " + q.first());
	}

}
